### ISSUE TITLE
USB device not bootable (syslinux fails)

### ISSUE THREAD

**Issue by jkirk** [grml/grml2usb#33]
Not sure yet what is going on, but syslinux (in grml2usb) fails to install the boot loader on my USB device:
```
% sudo ./grml2usb --fat16 /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1
Executing grml2usb version v0.18.0-2-g17293ed (git)
Are you sure you want to format the specified partition with fat16? y/N y
Note: you can skip this question using the option --force
Formating partition with fat16 filesystem
mkfs.fat 4.1 (2017-01-24)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Checking for boot flag
Identified grml flavour "grml64-full_testing".
Copying files. This might take a while....
No /EFI/ubuntu/grub.cfg found inside EFI image, looks like Secure Boot support is missing.
Synching data (this might take a while)
Installing default MBR
Installing syslinux as bootloader
syslinux: /dev/sdb1: No such file or directory
Fatal: Error executing syslinux (either try --fat16 or use grub?)
Cleaning up before exiting...
sudo ./grml2usb --fat16 /srv/data/isos/grml64-full_testing_build3133.iso   7.69s user 5.15s system 5% cpu 3:57.13 total
```

Running verbose (and only installing the bootloader) looks like this:
```
% sudo ./grml2usb --verbose --bootloader-only /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1                                                                                                                                                       
2020-05-30 17:27:05,770 Executing grml2usb version v0.18.0-2-g17293ed (git)                                                                                                                   
2020-05-30 17:27:05,841 Using ISO /srv/data/isos/grml64-full_testing_build3133.iso                                                                                                            
2020-05-30 17:27:05,842 mount ['-o', 'loop,ro', '-t', 'iso9660'] /srv/data/isos/grml64-full_testing_build3133.iso /tmp/grml2usbyu06831f                                                       
2020-05-30 17:27:05,882 register_mountpoint(/tmp/grml2usbyu06831f)                                                                                                                            
2020-05-30 17:27:05,890 Checking for boot flag                                                                                                                                                
2020-05-30 17:27:05,943 boot flag is enabled on /dev/sdb1                                                                                                                                     
2020-05-30 17:27:05,944 mount ['-o', 'utf8,iocharset=iso8859-1'] /dev/sdb1 /tmp/grml2usbq_brwekk                                                                                              
2020-05-30 17:27:06,131 register_mountpoint(/tmp/grml2usbq_brwekk)                                                                                                                            
2020-05-30 17:27:06,165 version_files = ['/tmp/grml2usbyu06831f/GRML/grml64-full_testing/grml-version']                                                                                       
2020-05-30 17:27:06,166 Identified grml flavour "grml64-full_testing".                                                                                                                        
2020-05-30 17:27:06,189 Copying addon file allinone.img                                                                                                                                       
2020-05-30 17:27:06,189 Source: /tmp/grml2usbyu06831f/boot/addons/allinone.img / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                   
2020-05-30 17:27:06,203 Copying addon file balder10.imz                                                                                                                                       
2020-05-30 17:27:06,203 Source: /tmp/grml2usbyu06831f/boot/addons/balder10.imz / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                   
2020-05-30 17:27:06,215 Copying addon file bsd4grml                                                                                                                                           
2020-05-30 17:27:06,216 Source: /tmp/grml2usbyu06831f/boot/addons/bsd4grml / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                       
2020-05-30 17:27:06,229 Copying addon file chain.c32                                                                                                                                          
2020-05-30 17:27:06,229 Source: /tmp/grml2usbyu06831f/boot/addons/chain.c32 / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                      
2020-05-30 17:27:06,232 Copying addon file hdt.c32                                                                                                                                            
2020-05-30 17:27:06,232 Source: /tmp/grml2usbyu06831f/boot/addons/hdt.c32 / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                        
2020-05-30 17:27:06,236 Copying addon file ipxe.lkrn                                                                                                                                          
2020-05-30 17:27:06,236 Source: /tmp/grml2usbyu06831f/boot/addons/ipxe.lkrn / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                     
2020-05-30 17:27:06,242 Copying addon file mboot.c32                                                                                                                                         
2020-05-30 17:27:06,242 Source: /tmp/grml2usbyu06831f/boot/addons/mboot.c32 / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                     
2020-05-30 17:27:06,248 Copying addon file memdisk                                                                                                                                           
2020-05-30 17:27:06,248 Source: /tmp/grml2usbyu06831f/boot/addons/memdisk / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                       
2020-05-30 17:27:06,251 Copying addon file memtest                                                                                                                                           
2020-05-30 17:27:06,251 Source: /tmp/grml2usbyu06831f/boot/addons/memtest / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                       
2020-05-30 17:27:06,256 Copying addon file menu.c32                                                                                                                                          
2020-05-30 17:27:06,257 Source: /tmp/grml2usbyu06831f/boot/addons/menu.c32 / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                      
2020-05-30 17:27:06,262 Copying addon file netboot.xyz.lkrn                                                                                                                                  
2020-05-30 17:27:06,262 Source: /tmp/grml2usbyu06831f/boot/addons/netboot.xyz.lkrn / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                              
2020-05-30 17:27:06,267 Copying addon file pci.ids                                                                                                                                           
2020-05-30 17:27:06,267 Source: /tmp/grml2usbyu06831f/boot/addons/pci.ids / Target: /tmp/grml2usbq_brwekk/boot/addons/                                                                       
2020-05-30 17:27:06,292 Source: /tmp/grml2usbyu06831f/boot/isolinux/logo.16 / Target: /tmp/grml2usbq_brwekk/boot/syslinux/logo.16                                                            
2020-05-30 17:27:06,307 Source: /tmp/grml2usbyu06831f/EFI/BOOT/bootx64.efi / Target: /tmp/grml2usbq_brwekk/efi/boot/bootx64.efi                                                              
2020-05-30 17:27:06,313 Source: /tmp/grml2usbyu06831f/boot/efi.img / Target: /tmp/grml2usbq_brwekk/boot/efi.img                                                                              
2020-05-30 17:27:06,316 efi_mountpoint = /tmp/grml2usbs9h25_gn                                                                                                                               
2020-05-30 17:27:06,316 mount(/tmp/grml2usbyu06831f/boot/efi.img, /tmp/grml2usbs9h25_gn, ['-o', 'ro', '-t', 'vfat']                                                                          
2020-05-30 17:27:06,317 mount ['-o', 'ro', '-t', 'vfat'] /tmp/grml2usbyu06831f/boot/efi.img /tmp/grml2usbs9h25_gn                                                                            
2020-05-30 17:27:06,394 register_mountpoint(/tmp/grml2usbs9h25_gn)                                                                                                                           
2020-05-30 17:27:06,395 ubuntu_cfg = None                                                                                                                                                    
2020-05-30 17:27:06,395 No /EFI/ubuntu/grub.cfg found inside EFI image, looks like Secure Boot support is missing.                                                                           
2020-05-30 17:27:06,396 umount [] /tmp/grml2usbs9h25_gn                                                                                                                                      
2020-05-30 17:27:06,442 unregister_mountpoint(/tmp/grml2usbs9h25_gn)                                                                                                                         
2020-05-30 17:27:06,442 Unmounted /tmp/grml2usbs9h25_gn                                                                                                                                      
2020-05-30 17:27:06,443 Removed directory /tmp/grml2usbs9h25_gn                                                                                                                              
2020-05-30 17:27:06,451 Source: /tmp/grml2usbyu06831f/boot/isolinux/f1 / Target: /tmp/grml2usbq_brwekk/boot/syslinux/f1  
[...]
2020-05-30 17:27:07,849 Found source file/boot/grub/grml64full_testing_default.cfg
2020-05-30 17:27:07,849 Found source file/boot/grub/grml64full_testing_options.cfg
2020-05-30 17:27:07,853 Generating syslinux configuration
2020-05-30 17:27:07,862 Updating grub configuration
2020-05-30 17:27:07,866 Synching data (this might take a while)
2020-05-30 17:28:39,126 umount [] /tmp/grml2usbq_brwekk
2020-05-30 17:28:39,386 unregister_mountpoint(/tmp/grml2usbq_brwekk)
2020-05-30 17:28:39,389 umount [] /tmp/grml2usbyu06831f
2020-05-30 17:28:39,446 unregister_mountpoint(/tmp/grml2usbyu06831f)
2020-05-30 17:28:39,447 Installing default MBR
2020-05-30 17:28:39,448 executing: dd if='/dev/sdb' of='/tmp/tmpnr8eh30m' bs=512 count=1
2020-05-30 17:28:39,673 executing: dd if=/home/jkirk/projects/grml/grml2usb/mbr/mbrldr of=/tmp/tmpnr8eh30m bs=439 count=1 conv=notrunc
2020-05-30 17:28:39,680 executing: dd if='/tmp/tmpnr8eh30m' of='/dev/sdb' bs=512 count=1 conv=notrunc
2020-05-30 17:28:39,689 executing: sync
2020-05-30 17:28:39,712 Installing syslinux as bootloader
2020-05-30 17:28:39,713 syslinux -d boot/syslinux /dev/sdb1
syslinux: /dev/sdb1: No such file or directory
2020-05-30 17:28:39,715 Fatal: Error executing syslinux (either try --fat16 or use grub?)
2020-05-30 17:28:39,715 Cleaning up before exiting...
2020-05-30 17:28:39,721 RuntimeError while removing temporary /tmp/grml2usbs9h25_gn, ignoring
```

`/dev/sdb1` exists and running syslinux manually works:
```
% file /dev/sdb1       
/dev/sdb1: block special (8/17)
% sudo syslinux -d boot/syslinux /dev/sdb1
```

**Comment by mika** [comment]
Hm, is there any hint in dmesg or syslog related to `/dev/sdb*`?

**Comment by jkirk** [comment]
Seems like something happens when installing the MBR:
```
% sudo ./grml2usb --skip-addons --bootloader-only /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1                             
[sudo] password for jkirk:
Executing grml2usb version v0.18.0-2-g17293ed-dirty (git)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Checking for boot flag
Identified grml flavour "grml64-full_testing".
Copying bootloader files (syslinux logo, EFI)...
No /EFI/ubuntu/grub.cfg found inside EFI image, looks like Secure Boot support is missing.
Copying bootloader files (isolinux)...
Copying bootloader files (syslinux modules)...
Copying bootloader files (syslinux addons)...
Copying bootloader files (grub)...
Synching data (this might take a while)
Check for /dev/sdb
brw-rw---- 1 root disk 8, 16 May 30 22:18 /dev/sdb
Installing default MBR
Check for /dev/sdb
ls: cannot access '/dev/sdb1': No such file or directory
Installing syslinux as bootloader
syslinux: /dev/sdb1: No such file or directory
Fatal: Error executing syslinux (either try --fat16 or use grub?)
Cleaning up before exiting...
```
The (relevant) logs in the meanwhile:
```
% sudo journalctl -f
May 30 22:18:22 executor sudo[881]:    jkirk : TTY=pts/2 ; PWD=/home/jkirk/projects/grml/grml2usb ; USER=root ; COMMAND=./grml2usb --skip-addons --bootloader-only /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1
May 30 22:18:22 executor sudo[881]: pam_unix(sudo:session): session opened for user root by jkirk(uid=0)                                                                                     
May 30 22:18:23 executor kernel: ISO 9660 Extensions: Microsoft Joliet Level 3
May 30 22:18:23 executor kernel: ISO 9660 Extensions: RRIP_1991A
May 30 22:18:23 executor systemd[1]: tmp-grml2usbi8adycek.mount: Succeeded.
May 30 22:18:23 executor systemd[4066]: tmp-grml2usbi8adycek.mount: Succeeded.
May 30 22:19:52 executor systemd[4066]: tmp-grml2usb_sq9vrfw.mount: Succeeded.
May 30 22:19:52 executor systemd[1]: tmp-grml2usb_sq9vrfw.mount: Succeeded.
May 30 22:19:52 executor systemd[1]: tmp-grml2usby2fgg8om.mount: Succeeded.
May 30 22:19:52 executor systemd[4066]: tmp-grml2usby2fgg8om.mount: Succeeded.
May 30 22:19:53 executor sudo[881]: pam_unix(sudo:session): session closed for user root
May 30 22:19:54 executor kernel:  sdb: sdb1
```

This is how it looks like if I add `--skip-mbr`:
```
% sudo ./grml2usb --skip-addons --skip-mbr --bootloader-only /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1
Executing grml2usb version v0.18.0-2-g17293ed-dirty (git)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Checking for boot flag
Identified grml flavour "grml64-full_testing".
Copying bootloader files (syslinux logo, EFI)...
No /EFI/ubuntu/grub.cfg found inside EFI image, looks like Secure Boot support is missing.
Copying bootloader files (isolinux)...
Copying bootloader files (syslinux modules)...
Copying bootloader files (syslinux addons)...
Copying bootloader files (grub)...
Synching data (this might take a while)
Check for /dev/sdb*
brw-rw---- 1 root disk 8, 17 May 30 22:24 /dev/sdb1
Installing syslinux as bootloader
Note: grml flavour grml64-full_testing was installed as the default booting system.
Note: you can boot flavour grml64-full_testing using 'grml64-full_testing' on the commandline.
Finished execution of grml2usb (v0.18.0-2-g17293ed-dirty (git)). Have fun with your Grml system.

May 30 22:24:00 executor sudo[2968]:    jkirk : TTY=pts/2 ; PWD=/home/jkirk/projects/grml/grml2usb ; USER=root ; COMMAND=./grml2usb --skip-addons --skip-mbr --bootloader-only /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1
May 30 22:24:00 executor sudo[2968]: pam_unix(sudo:session): session opened for user root by jkirk(uid=0)
May 30 22:24:00 executor kernel: ISO 9660 Extensions: Microsoft Joliet Level 3
May 30 22:24:00 executor kernel: ISO 9660 Extensions: RRIP_1991A
May 30 22:24:01 executor systemd[1]: tmp-grml2usbe9nm9gyf.mount: Succeeded.
May 30 22:24:01 executor systemd[4066]: tmp-grml2usbe9nm9gyf.mount: Succeeded.
May 30 22:24:16 executor sudo[3650]:    jkirk : TTY=pts/17 ; PWD=/home/jkirk ; USER=root ; COMMAND=/bin/journalctl -f
May 30 22:24:16 executor sudo[3650]: pam_unix(sudo:session): session opened for user root by jkirk(uid=0)
May 30 22:25:29 executor systemd[1]: tmp-grml2usb459c7sik.mount: Succeeded.
May 30 22:25:29 executor systemd[4066]: tmp-grml2usb459c7sik.mount: Succeeded.
May 30 22:25:30 executor systemd[1]: tmp-grml2usbtdmmiied.mount: Succeeded.
May 30 22:25:30 executor systemd[4066]: tmp-grml2usbtdmmiied.mount: Succeeded.
May 30 22:25:31 executor sudo[2968]: pam_unix(sudo:session): session closed for user root
```
```

**Comment by jkirk** [comment]
I seems like a kind of partprobe is needed before installing syslinux. Adding `check_boot_flag(device)` right before installing syslinux fixes the problem for me.

Checking the boot flag (again) seems not nice[tm] but maybe running something like [parted.getDevice(boot_dev)](https://github.com/grml/grml2usb/blob/master/grml2usb#L500) should be enough? Any thoughts?

**Comment by mika** [comment]
Is the `parted.getDevice(boot_dev)` enough to get it working for you?

I didn't look closer into it, but `_ped` (the underlying libparted binding of python3-parted) has a:

```
from _ped import device_probe_all
device_probe_all()
```

Does any of this adding it to the end of the mbr code help?

**Comment by jkirk** [comment]
Wow, this drove me nearly nuts :exploding_head: 

I tried all of the following (at the end of `install_mbr`):
```
  import _ped
  _ped.device_get(device)
```
```
  import _ped
  _ped.device_probe_all
```
```
  import parted
  parted.getDevice(device)
```
(Very) sometimes it worked, most of the time it didn't.

Then I tried this:
```
  logging.info("Probing device...")
  proc = subprocess.Popen(["partprobe", device])
```
which worked better[tm] (`Check for /dev/sdb` failed sometimes, but at least `syslinux` could be installed nearly every time).
And finally I tried this, which I found worked most reliably:
```
  logging.info("Probing device...")
  proc = subprocess.Popen(["blockdev", "--rereadpt", "/dev/sdb"])
```
The problem seems to be that, the `libparted` functions seem to get the information they need, but they seem to be non-blocking and there seems to be a time delay till the kernel gets informed, all udev, etc.. rules are ran through and the block device path `/dev/sdb1` is re-created. I could not figure out how to make sure "everything is settled" after the kernel has been informed to re-read the partition table...

A few questions came up while looking at the code:
* I added some `logging.info` lines:
```
Copying bootloader files (syslinux logo, EFI)...
Copying bootloader files (isolinux)...
Copying bootloader files (syslinux modules)...
Copying bootloader files (syslinux addons)...
Copying bootloader files (grub)...
```
Shall we keep them, remove them or shall I make them `logging.debug`?
* We seem to activate the partition in https://github.com/grml/grml2usb/blob/master/grml2usb#L833. Why do we then check the boot flag if we _try_ to set it anyway? (Just checked with `--skip-bootflag` after removing the boot flag and it was **not** set after the `grml2usb` run. So this doesn't seem to work anyway.)
* The comment for `@mbrtemplate` is given twice.
